Mathlib/GroupTheory/DivisibleHull.lean

```diff
@@ -0,0 +1,405 @@
+  simp [qsmul_def]
+
+theorem qsmul_of_nonneg {a : ℚ} (h : 0 ≤ a) (x : DivisibleHull M) :
+    a • x = (show ℚ≥0 from ⟨a, h⟩) • x := by
+  have := h.eq_or_lt
+  aesop (add simp [qsmul_def, abs_of_pos])
+
+theorem qsmul_of_nonpos {a : ℚ} (h : a ≤ 0) (x : DivisibleHull M) :
+    a • x = -((show ℚ≥0 from ⟨-a, Left.nonneg_neg_iff.mpr h⟩) • x) := by
+  have := h.eq_or_lt
+  aesop (add simp [qsmul_def, abs_of_neg])
+
+theorem qsmul_mk (a : ℚ) (m : M) (s : ℕ+) :
+    a • mk m s = mk (a.num • m) (⟨a.den, a.den_pos⟩ * s) := by
+  obtain h | h := le_total 0 a
+  · rw [qsmul_of_nonneg h, nnqsmul_mk, ← natCast_zsmul]
+    congr
+    simpa using h
+  · rw [qsmul_of_nonpos h]
+    have : a.num.natAbs • m = -a.num • m := by
+      rw [← natCast_zsmul]
+      congr
+      simpa using h
+    simp [nnqsmul_mk, this, ← neg_mk]
+
+noncomputable
+instance : Module ℚ (DivisibleHull M) where
+  one_smul x := by
+    induction x with | mk m s
+    simp [qsmul_of_nonneg zero_le_one, nnqsmul_mk]
+  zero_smul := zero_qsmul
+  smul_zero a := by simp [qsmul_def]
+  smul_add a x y := by simp [qsmul_def, smul_add]
+  add_smul a b x := by
+    induction x with | mk m s
+    simp_rw [qsmul_mk, mk_add_mk, mk_eq_mk]
+    use 1
+    suffices ((a + b).num * a.den * b.den * (s * s)) • m =
+        ((a.num * b.den + b.num * a.den) * (a + b).den * (s * s)) • m  by
+      convert this using 1
+      all_goals
+      simp [← natCast_zsmul, smul_smul, ← add_smul]
+      ring_nf
+    rw [Rat.add_num_den']
+  mul_smul a b x := by
+    induction x with | mk m s
+    simp_rw [qsmul_mk, mk_eq_mk]
+    use 1
+    suffices ((a * b).num * a.den * b.den * s) • m = (a.num * b.num * (a * b).den * s) • m by
+      convert this using 1
+      all_goals
+      simp [← natCast_zsmul, smul_smul]
+      ring_nf
+    rw [Rat.mul_num_den']
+
+theorem zsmul_mk (a : ℤ) (m : M) (s : ℕ+) : a • mk m s = mk (a • m) s := by
+  simp [← Int.cast_smul_eq_zsmul ℚ a, qsmul_mk]
+
+end Group
+
+section LinearOrder
+variable {M : Type*} [AddCommMonoid M] [LinearOrder M] [IsOrderedCancelAddMonoid M]
+
+private theorem lift_aux (m n m' n' : M) (s t s' t' : ℕ+)
+    (h : mk m s = mk m' s') (h' : mk n t = mk n' t') :
+    (t.val • m ≤ s.val • n) = (t'.val • m' ≤ s'.val • n') := by
+  rw [mk_eq_mk_iff_smul_eq_smul] at h h'
+  rw [propext_iff, ← (nsmul_right_strictMono (mul_ne_zero s'.ne_zero t'.ne_zero)).le_iff_le]
+  convert (nsmul_right_strictMono (M := M) (mul_ne_zero s.ne_zero t.ne_zero)).le_iff_le using 2
+  · simp_rw [smul_smul, mul_rotate s'.val, ← smul_smul, h, smul_smul]
+    ring_nf
+  · simp_rw [smul_smul, ← mul_rotate s'.val, ← smul_smul, ← h', smul_smul]
+    ring_nf
+instance : LE (DivisibleHull M) where
+  le x y := liftOn₂ x y (fun m s n t ↦ t.val • m ≤ s.val • n) lift_aux
+
+theorem mk_le_mk {m m' : M} {s s' : ℕ+} :
+    mk m s ≤ mk m' s' ↔ s'.val • m ≤ s.val • m' := by rfl
+
+instance : LinearOrder (DivisibleHull M) where
+  le_refl a := by
+    induction a with | mk m s
```

Comment:
```suggestion
  le_refl
  | mk m s =>
```
should work

Reply:
It didn't work
```
Invalid pattern: Expected a constructor or constant marked with `[match_pattern]`

Hint: These are similar:
  'Add.mk',
  'AddAction.IsPretransitive.mk',
  'AddAction.mk',
  'AddActionHom.mk',
  'AddActionSemiHomClass.mk',
  'AddCancelCommMonoid.mk',
  'AddCancelMonoid.mk',
  'AddCommGroup.mk',
  'AddCommGroupWithOne.mk',
  'AddCommMagma.mk',
```
I also tried tagging `ind` with `match_pattern` but it doesn't allow me